### ISSUE TITLE
I need sublevel more

### ISSUE THREAD

**Issue by MichaelLiew1990** [shun126/UE5-DungeonGeneratorDemo#53]
I want spawn a 10x10 sublevel for my boss place, but i don't want to change other random room size (random room size is 2-6), this boss room spawn using a probability to control, and there will exists multi boss rooms with different settings

**Comment by shun126** [comment]
Thank you for contacting me.

I assumed the goal room would be the boss room.

Is the boss room the goal room? If it is always drawn to the regular room, would it serve the purpose?
I imagine it like the middle boss room that always appears.

**Comment by MichaelLiew1990** [comment]
In my game, goal room is boss room, but i need some big rooms in the middle, these rooms serve as amazing custom room

**Comment by MichaelLiew1990** [comment]
Hi, Do you have any plans to add this feature? If so, when will it be completed? Otherwise, I will make the modifications of source code myself. Additionally, here is another bug show in pic.
![image](https://github.com/user-attachments/assets/f025d156-d60d-492a-a874-49a3459552ce)

**Comment by shun126** [comment]
I'm truly sorry for the inconvenience. We are currently addressing the issue of package failure. I will start working on this issue after that.

**Comment by shun126** [comment]
believe I have figured out the issue with the rooms overlapping. There have been reports of replication not working on Steam, so I will consider fixing this at the same time.

**Comment by shun126** [comment]
![preferred_](https://github.com/user-attachments/assets/ef4957b4-57f5-4b1e-9007-c18a201bc7ed)

I’m thinking of adding a system to prioritize the placement of sublevels, separate from the existing system where sublevels are drawn for rooms. There are cases where the placement instructions from the Mission Graph, such as placing keys, do not match, so these sublevels will be excluded from the key placement candidates.

Sorry for the separate question, but have you ever encountered issues when implementing matchmaking through Steam?

**Comment by shun126** [comment]
I have submitted version 1.6.7 to Epic, but it seems the review process is taking longer than expected.

**Comment by shun126** [comment]
I sincerely apologize for the wait. Version 1.6.9 has been approved. Please update the plugin.

**Comment by MichaelLiew1990** [comment]
Show like pic below, i update from 1.6.5 to 1.6.10 for testing, and set up a preferred sublevel, and i found a weird random sublevel room, it size big than max room size
![BUG1](https://github.com/user-attachments/assets/235ba64d-0110-4147-82a2-a43d2b572c96)

**Comment by shun126** [comment]
By any chance, is there a 'Expanded the room because it was too small for the connecting gate' message in the Verbose log? If it's displayed, the process of expanding the room when it's too small for the connecting gate is affecting the issue.

I will investigate whether there are any other possible causes and look for a workaround.

**Comment by MichaelLiew1990** [comment]
> By any chance, is there a 'Expanded the room because it was too small for the connecting gate' message in the Verbose log? If it's displayed, the process of expanding the room when it's too small for the connecting gate is affecting the issue.
> 
> I will investigate whether there are any other possible causes and look for a workaround.

Not output that message, I think is goal map problem, you can see this pic
![image](https://github.com/user-attachments/assets/8666fae0-e4c2-46d6-be96-34da3e75907e)

**Comment by shun126** [comment]
Thank you. I think I’ve identified the cause. Due to the transition of the Marketplace to Fab, the update timeline is unclear. I apologize if you are in a hurry.

**Comment by MichaelLiew1990** [comment]
> Thank you. I think I’ve identified the cause. Due to the transition of the Marketplace to Fab, the update timeline is unclear. I apologize if you are in a hurry.

Yes, i am in a hurry now, if Fab will wait for long time, you can send the source code directly to my e-mail michaelliew@126.com, very thanks!

**Comment by MichaelLiew1990** [comment]
Hi, I updated to 1.6.11, and saw your modify:
			// 部屋の最大サイズにあわせる
			if (width > mGenerateParameter.GetMaxRoomWidth())
				width = mGenerateParameter.GetMaxRoomWidth();
			if (depth > mGenerateParameter.GetMaxRoomDepth())
				depth = mGenerateParameter.GetMaxRoomDepth();
This only solved Goal Room problem, but Preferred SubLevel still have this problem

**Comment by shun126** [comment]
It might be fixed if you change it as follows. The issue with errors occurring during sub-level loading is currently under investigation.

```cpp
// 部屋の最大サイズにあわせる
if (width > mGenerateParameter.GetMaxRoomWidth())
{
	width = mGenerateParameter.GetMaxRoomWidth();
	room->SetWidth(width);
}
if (depth > mGenerateParameter.GetMaxRoomDepth())
{
	depth = mGenerateParameter.GetMaxRoomDepth();
	room->SetDepth(depth);
}
if (height > mGenerateParameter.GetMaxRoomHeight())
{
	height = mGenerateParameter.GetMaxRoomHeight();
	room->SetHeight(depth);
}
```

**Comment by shun126** [comment]
Sorry, is there a way to reproduce the LoadError?

**Comment by MichaelLiew1990** [comment]
> Sorry, is there a way to reproduce the LoadError?

Thanks a lot, modify code like than can solve max room problem.
I don't know how to reproduce the LoadError, it just occur sometime, I copy the seed when occur, and using that seed to reproduce, but it no error. Can you make some work to walkaround it? not let everything go wrong when LoadError.

Another suggest: Preferred SubLevel Pool should set a max num, for example, i set 5 SubLevels in the pool and set the weight for everyone, then set max num to 2, it will select and load 2 Preferred SubLevels at runtime

**Comment by shun126** [comment]
Looking at the reason for the Failed to Load Level error, it's possible that the level could not be unloaded or wasn't unloaded in time.

I think setting a maximum number for sub-level selection is a good idea. I'll consider it.

**Comment by shun126** [comment]
I have created version 1.6.12 beta. Hopefully, the sub-level release issue has been fixed.
You should be able to use it by extracting the zip file into the Plugin directory of your project directory

https://www.dropbox.com/scl/fo/9k7r7nso6wewb70c8o6j2/AIkLnP2GTotzRx9AW4czcxk?rlkey=ok7adc5oh7mhqypqewbpc6jw1&st=8w0rx38z&dl=0

**Comment by MichaelLiew1990** [comment]
> I have created version 1.6.12 beta. Hopefully, the sub-level release issue has been fixed. You should be able to use it by extracting the zip file into the Plugin directory of your project directory
> 
> https://www.dropbox.com/scl/fo/9k7r7nso6wewb70c8o6j2/AIkLnP2GTotzRx9AW4czcxk?rlkey=ok7adc5oh7mhqypqewbpc6jw1&st=8w0rx38z&dl=0

OK, Thanks, I have downloaded, you can delete this share, cause your plugin is not free for everyone

**Comment by shun126** [comment]
I’m glad you were able to download it. Thank you for letting me know.

This is a beta version. Due to a structural change midway, the member functions in the dungeon generation-related classes haven’t been fully organized. I’ll arrange them in the future.

Yes, any user can download it. So if you are satisfied with it,please consider purchasing it to support my activities. :-)

**Comment by MichaelLiew1990** [comment]
> I’m glad you were able to download it. Thank you for letting me know.
> 
> This is a beta version. Due to a structural change midway, the member functions in the dungeon generation-related classes haven’t been fully organized. I’ll arrange them in the future.
> 
> Yes, any user can download it. So if you are satisfied with it,please consider purchasing it to support my activities. :-)

I have purchased few month ago :-)  I have tested 1.6.12 today, SubLevel Error has gone. but your code have some warning:
1.
void ADungeonGenerateActor::PostInitializeComponents()
{
	// Calling the parent class
	Super::PostInitializeComponents();

	if (GetNetMode() != NM_Standalone)
	{
		SetReplicates(true);   //Here will report warning
		SetAutonomousProxy(true);   //Here will report warning
	}
}

2.Everything like this meta = (AllowedClasses = "Actor") should change to "/Script/Engine.Actor"

**Comment by shun126** [comment]
The beta version has been deleted now that the operation has been confirmed. Thank you very much for your cooperation with feature ideas and bug fixes.

The first warning was triggered when the client modified the replication, so I was able to fix it by changing if (GetNetMode() != NM_Standalone) to if (GetLocalRole() == ROLE_Authority).

I don’t fully understand the meaning of the second warning. Could you explain what it’s about?
